packages/design-system/src/components/FieldContainer/FieldContainer.jsx

```diff
@@ -0,0 +1,165 @@
+import { omit, uniqueId } from 'lodash';
+import FormLabel from '../FormLabel/FormLabel';
+import PropTypes from 'prop-types';
+import React from 'react';
+import classNames from 'classnames';
+
+// Props that are exclusive used by <FieldContainer> in a form field component
+// These shouldn't be passed to the field input elements, i.e. <Select> or <TextInput>
+// Certain props (`id`, `labelId`, `inputRef`, `focusTrigger) will be transformed/renamed
+// and then forwarded to the field input  elements via `fieldProps`
+export const containerPropList = [
+  'className',
+  'errorMessage',
+  'fieldClassName',
+  // TODO: rename to `autoFocus`
+  'focusTrigger',
+  'hint',
+  // TODO: consider renaming this to `fieldRef`
+  'id',
+  // TODO: consider rename this to `fieldRef`
+  'inputRef',
+  'inversed',
+  'label',
+  'labelClassName',
+  'labelId',
+  'name',
+  'requirementLabel',
+];
+
+export class FieldContainer extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.id = props.id || uniqueId(`${this.props.name || 'field'}_`);
+    this.labelId = props.labelId || uniqueId(`${this.props.name || 'field'}_label_`);
+    this.setFieldRef = this.setFieldRef.bind(this);
+  }
+
+  componentDidMount() {
+    if (this.props.focusTrigger) {
+      this.focusRef && this.focusRef.focus();
+    }
+  }
+
+  setFieldRef(elem) {
+    // Use React.forwardRef when upgraded to React 16.3
+    if (this.props.focusTrigger) {
+      this.focusRef = elem;
+    }
+    if (this.props.inputRef) {
+      this.props.inputRef(elem);
+    }
+  }
+
+  render() {
+    const containerProps = omit(this.props, containerPropList);
+
+    const ComponentType = this.props.component;
+    const classes =
+      ComponentType === 'fieldset'
+        ? classNames(this.props.className, 'ds-c-fieldset')
+        : this.props.className;
+
+    // Props that have been transformed or renamed by <FieldContainer>
+    const transformedFieldProps = {
+      className: this.props.fieldClassName,
```

Comment:
Because our existing form components `props` along to different components, our we have prefixed props. I.e. `TextField` accepts `className`, `labelClassName` and `fieldClassName`, which are all applied to the root `div` element, the `label` element, and the `input` element respectively. When creating a `TextInput` element, it would be ideal to use a `className` prop. However, this introduces an issue where `FieldContainer` has to map `fieldClassName` to `className` when it is passed to the field input.

Reply:
The drawback is that you would have to understand how `FieldContainer` renames these props. Plus, most field input components like `TextInput` and `Select` will be private anyway, so maybe the consistency of the prop naming convention isnt as important